### PR TITLE
Native tokens

### PR DESCRIPTION
**Pull request by ChristianoBraga**
Add trivial native token policy contract.

### CODE REVIEW

**Review by SchmErik — CHANGES_REQUESTED** [review]
I think this looks good but we really need to remove the 2 unintended commits as they're adding changes. You left directions on how to build the tests. Could you copy this to a readme file inside of native tokens directory and also mention the commit id of the relevant plutus repos that you used to test this?

File: native-tokens/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# How to generate a trivial native-tokens policy
+
+1. Install IOG's `plutus`. After installing, run `nix-shell` at the
+plutus directory.
+   - Plutus commit ID: bcdd1ceffc93c733d90f5b89bcfae47d21dc2fcd.
+2. Move to `plutus-semantics/native-tokens/native-tokens`.
+3. Run `cabal build`.
+4. Run `cabal run native-tokens-script out_scripts scritps`.
```
Comment (SchmErik) [inline native-tokens/README.md:8]: did you mean scripts`.?

File: native-tokens/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# How to generate a trivial native-tokens policy
+
+1. Install IOG's `plutus`. After installing, run `nix-shell` at the
```
Comment (SchmErik) [inline native-tokens/README.md:3]: What do you mean by installing IOG's plutus? Do you mean "clone IoG's Plutus repo"? If so, you may want to change the wording. Installing and cloning are different things

**Review by SchmErik — CHANGES_REQUESTED** [review]

File: native-tokens/README.md (hunk 2 above)
Comment (ChristianoBraga) [reply native-tokens/README.md:3]: Install.

File: native-tokens/README.md (hunk 1 above)
Comment (ChristianoBraga) [reply native-tokens/README.md]: Done

**Comment by SchmErik** [comment]
I'm getting this message:

```
Failed to build lzma-0.0.0.3.
Build log (
/root/.cabal/logs/ghc-8.10.7/lzma-0.0.0.3-b8e5e5271afb9de54972ab2bbbbfe6e55609cefdde55c24b9e9d56d7e4cf3ba8.log
):
Configuring library for lzma-0.0.0.3..
Preprocessing library for lzma-0.0.0.3..
compiling dist/build/LibLzma_hsc_make.c failed (exit code 1)
rsp file was: "dist/build/hsc2hscall1477540-0.rsp"
command was: /nix/store/pja9g36cy32z3d51942jqk91a6l2d5nv-gcc-wrapper-10.3.0/bin/cc -c dist/build/LibLzma_hsc_make.c -o dist/build/LibLzma_hsc_make.o -D__GLASGOW_HASKELL__=810 -Dlinux_BUILD_OS=1 -Dx86_64_BUILD_ARCH=1 -Dlinux_HOST_OS=1 -Dx86_64_HOST_ARCH=1 -Idist/build/autogen -Idist/build/global-autogen -include dist/build/autogen/cabal_macros.h -I/nix/store/ngzxn0xjzh3xc4x7kr9s3wp74qkfrvms-ghc-8.10.7/lib/ghc-8.10.7/bytestring-0.10.12.0/include -I/nix/store/ngzxn0xjzh3xc4x7kr9s3wp74qkfrvms-ghc-8.10.7/lib/ghc-8.10.7/base-4.14.3.0/include -I/nix/store/ryschvm8md9iswan8g0azdgaxpqqs4m5-gmp-6.2.1-dev/include -I/nix/store/ngzxn0xjzh3xc4x7kr9s3wp74qkfrvms-ghc-8.10.7/lib/ghc-8.10.7/integer-gmp-1.0.3.0/include -I/nix/store/cljs2lqbh64iwnpfmgc8kz2w1zbn2r4a-numactl-2.0.14/include -I/nix/store/ngzxn0xjzh3xc4x7kr9s3wp74qkfrvms-ghc-8.10.7/lib/ghc-8.10.7/include -I/nix/store/bcif7a6m44x6dqq2fgxiih2ags0hqqnf-libffi-3.4.2-dev/include -I/nix/store/ngzxn0xjzh3xc4x7kr9s3wp74qkfrvms-ghc-8.10.7/lib/ghc-8.10.7/include/
error: LibLzma.hsc:44:10: fatal error: lzma.h: No such file or directory
compilation terminated.

cabal: Failed to build lzma-0.0.0.3 (which is required by
exe:native-tokens-scripts from native-tokens-0.1.0.0). See the build log above
for details.
```
@ChristianoBraga I remember you saying that you were able to resolve this. Could you add directions on how to resolve this message?

**Comment by ChristianoBraga** [comment]
> I added this change to shell.nix in the plutus repo and ran the build again. This time, it got stuck here:
> 
> ```
> [ 9 of 20] Compiling Database.Beam.Migrate.SQL.Tables ( Database/Beam/Migrate/SQL/Tables.hs, dist/build/Database/Beam/Migrate/SQL/Tables.o, dist/build/Database/Beam/Migrate/SQL/Tables.dyn_o )
> 
> Database/Beam/Migrate/SQL/Tables.hs:32:5: warning: [-Wdodgy-exports]
>     The export item ‘IsNotNull(..)’ suggests that
>     ‘IsNotNull’ has (in-scope) constructors or class methods,
>     but it has none
>    |
> 32 |   , IsNotNull(..)
>    |     ^^^^^^^^^^^^^
> [10 of 20] Compiling Database.Beam.Migrate.SQL.BeamExtensions ( Database/Beam/Migrate/SQL/BeamExtensions.hs, dist/build/Database/Beam/Migrate/SQL/BeamExtensions.o, dist/build/Database/Beam/Migrate/SQL/BeamExtensions.dyn_o )
> [11 of 20] Compiling Database.Beam.Migrate.SQL ( Database/Beam/Migrate/SQL.hs, dist/build/Database/Beam/Migrate/SQL.o, dist/build/Database/Beam/Migrate/SQL.dyn_o )
> [12 of 20] Compiling Database.Beam.Migrate.SQL.Builder ( Database/Beam/Migrate/SQL/Builder.hs, dist/build/Database/Beam/Migrate/SQL/Builder.o, dist/build/Database/Beam/Migrate/SQL/Builder.dyn_o )
> [13 of 20] Compiling Database.Beam.Migrate.Generics.Types ( Database/Beam/Migrate/Generics/Types.hs, dist/build/Database/Beam/Migrate/Generics/Types.o, dist/build/Database/Beam/Migrate/Generics/Types.dyn_o )
> [14 of 20] Compiling Database.Beam.Migrate.Generics ( Database/Beam/Migrate/Generics.hs, dist/build/Database/Beam/Migrate/Generics.o, dist/build/Database/Beam/Migrate/Generics.dyn_o )
> [15 of 20] Compiling Database.Beam.Haskell.Syntax ( Database/Beam/Haskell/Syntax.hs, dist/build/Database/Beam/Haskell/Syntax.o, dist/build/Database/Beam/Haskell/Syntax.dyn_o )
> 
> Database/Beam/Haskell/Syntax.hs:1041:10: error:
>     Duplicate instance declarations:
>       instance Hashable a => Hashable (S.Set a)
>         -- Defined at Database/Beam/Haskell/Syntax.hs:1041:10
>       instance Hashable v => Hashable (S.Set v)
>         -- Defined in ‘hashable-1.3.5.0:Data.Hashable.Class’
>      |
> 1041 | instance Hashable a => Hashable (S.Set a) where
>      |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> cabal: Failed to build beam-migrate-0.5.1.0 (which is required by
> exe:native-tokens-scripts from native-tokens-0.1.0.0). See the build log above
> for details.
> ```

I have never run into this problem. This looks like a Haskell library. Maybe is important to use IOG's caches to build Nix not only to make the build faster but also to make sure Haskell has all the "necessary" libraries. There is a warning in Plutus doc saying that you should stop if you see Haskell being built. This is at the point of Plutus nix installation/configuration yes but that is the only thing that comes to my mind right now...

**Comment by SchmErik** [comment]
I installed the caches and i still get this error

**Comment by ChristianoBraga** [comment]
> Database.Beam.Migrate.SQL.Tables

@SchmErik I think this is related with your Haskell environment. I do not see what else I can say about this. Here are the versions of cabal and ghc I am using. I will keep investigating this but I think my instructions for creating the policy are accurate. I would like to close this for today's meeting.

```$ cabal --version
cabal-install version 3.6.2.0
compiled using version 3.6.2.0 of the Cabal library 
semantics:qua jul 20@10:39:34:plutus (plutus-core-specification-v3)$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.10.4.20210212
```

**Review by SchmErik — APPROVED** [review]

**Review by SchmErik — APPROVED** [review]

**Pull request merged**